src/factor_graph.cc

```diff
@@ -0,0 +1,250 @@
+  assert(capacity.num_variables == size.num_variables);
+  assert(capacity.num_factors == size.num_factors);
+  assert(capacity.num_edges == size.num_edges);
+  assert(capacity.num_weights == size.num_weights);
+
+  // check whether variables, factors, and weights are stored
+  // in the order of their id
+  for (variable_id_t i = 0; i < size.num_variables; ++i) {
+    assert(this->variables[i].id == i);
+  }
+  for (factor_id_t i = 0; i < size.num_factors; ++i) {
+    assert(this->factors[i].id == i);
+  }
+  for (weight_id_t i = 0; i < size.num_weights; ++i) {
+    assert(this->weights[i].id == i);
+  }
+}
+
+CompactFactorGraph::CompactFactorGraph(const FactorGraphDescriptor &size)
+    : size(size),
+      variables(new Variable[size.num_variables]),
+      factors(new Factor[size.num_factors]),
+      compact_factors(new CompactFactor[size.num_edges]),
+      compact_factors_weightids(new weight_id_t[size.num_edges]),
+      factor_ids(new factor_id_t[size.num_edges]),
+      vifs(new VariableInFactor[size.num_edges]) {}
+
+CompactFactorGraph::CompactFactorGraph(const CompactFactorGraph &other)
+    : CompactFactorGraph(other.size) {
+  // copy each member from the given graph
+  COPY_ARRAY_UNIQUE_PTR_MEMBER(variables, size.num_variables);
+  COPY_ARRAY_UNIQUE_PTR_MEMBER(factors, size.num_factors);
+  COPY_ARRAY_UNIQUE_PTR_MEMBER(compact_factors, size.num_edges);
+  COPY_ARRAY_UNIQUE_PTR_MEMBER(compact_factors_weightids, size.num_edges);
+  COPY_ARRAY_UNIQUE_PTR_MEMBER(factor_ids, size.num_edges);
+  COPY_ARRAY_UNIQUE_PTR_MEMBER(vifs, size.num_edges);
+}
+
+weight_id_t CompactFactorGraph::get_multinomial_weight_id(
+    const variable_value_t assignments[], const CompactFactor &fs,
+    variable_id_t vid, variable_value_t proposal) {
+  /**
+   * For FUNC_MULTINOMIAL, the weight ids are aligned in a continuous region
+   * according
+   * to the numerical order of variable values.
+   * For example, for variable assignment indexes i1, ..., ik with cardinality
+   * d1, ..., dk
+   * The weight index is
+   * (...((((0 * d1 + i1) * d2) + i2) * d3 + i3) * d4 + ...) * dk + ik
+   *
+   * For FUNC_SPARSE_MULTINOMIAL, we look up the weight_ids map.
+   *
+   * TODO: refactor the above formula into a shared routine. (See also
+   * binary_parser.read_factors)
+   */
+  weight_id_t weight_offset = 0;
+  // for each variable in the factor
+  for (size_t i = fs.n_start_i_vif; i < fs.n_start_i_vif + fs.n_variables;
+       ++i) {
+    const VariableInFactor &vif = vifs[i];
+    const Variable &variable = variables[vif.vid];
+    weight_offset *= variable.cardinality;
+    weight_offset += variable.get_domain_index(
+        vif.vid == vid ? proposal : assignments[vif.vid]);
+  }
+
+  switch (fs.func_id) {
+    case FUNC_SPARSE_MULTINOMIAL: {
+      auto iter = factors[fs.id].weight_ids->find(weight_offset);
+      if (iter != factors[fs.id].weight_ids->end()) return iter->second;
+      break;
+    }
+    case FUNC_MULTINOMIAL:
```

Comment:
I'd kill the code paths for this. The implicit weight parameterization never made any sense.